File: yapcore/src/main/java/co/yap/yapcore/helpers/validation/rule/TypeRule.java
package co.yap.yapcore.helpers.validation.rule;

import android.widget.TextView;

import androidx.annotation.StringRes;

import java.lang.reflect.InvocationTargetException;

import co.yap.yapcore.R;

/**
 * Created irfan arshad on 10/6/2020.
 */
public abstract class TypeRule extends Rule<TextView, TypeRule.FieldType> {

    public TypeRule(TextView view, FieldType value, String errorMessage) {
        super(view, value, errorMessage);
    }

    public enum FieldType {
        Cpf(CpfTypeRule.class, R.string.error_message_cpf_validation),
        Username(UsernameRule.class, R.string.error_message_username_validation),
        Email(EmailTypeRule.class, R.string.error_message_email_validation),
        Url(UrlTypeRule.class, R.string.error_message_url_validation),
        PostalAddress(PostalAddressRule.class, R.string.error_message_postal_address),
        None;

        public @StringRes
        int errorMessageId;
        Class<? extends TypeRule> mClass;

        FieldType(Class<? extends TypeRule> mClass, @StringRes int errorMessageId) {
            this.mClass = mClass;
            this.errorMessageId = errorMessageId;
        }

        FieldType() {
        }

        public TypeRule instantiate(TextView view, String errorMessage) throws NoSuchMethodException, IllegalAccessException
                , InvocationTargetException, InstantiationException {
            if (this != None) {
                return mClass.getConstructor(TextView.class, String.class).newInstance(view, errorMessage);
            }
            throw new IllegalStateException("It's not possible to bind a none value type");
        }
    }
}

File: yapcore/src/main/java/co/yap/yapcore/helpers/validation/rule/TypeRule.kt
package co.yap.yapcore.helpers.validation.rule

import android.widget.TextView
import androidx.annotation.Keep
import androidx.annotation.StringRes
import co.yap.yapcore.R
import java.lang.reflect.InvocationTargetException

/**
 * Created irfan arshad on 10/6/2020.
 */
@Keep
abstract class TypeRule(
    view: TextView?,
    value: FieldType?,
    errorMessage: String?,
    errorEnabled: Boolean
) : Rule<TextView?, TypeRule.FieldType?>(
    view,
    value,
    errorMessage,
    errorEnabled
) {
    enum class FieldType {
        Cpf(
            CpfTypeRule::class.java,
            R.string.error_message_cpf_validation
        ),
        FirstName(
            FirstNameRule::class.java, R.string.error_message_first_name_validation
        ),
        LastName(
            LastNameRule::class.java, R.string.error_message_last_name_validation
        ),
        Email(
            EmailTypeRule::class.java, R.string.error_message_email_validation
        ),
        Url(
            UrlTypeRule::class.java, R.string.error_message_url_validation
        ),
        PostalAddress(
            PostalAddressRule::class.java,
            R.string.error_message_credit_card_validation
        ),
        None;

        var mClass: Class<out TypeRule>? = null

        @JvmField
        @StringRes
        var errorMessageId = 0

        constructor(
            mClass: Class<out TypeRule>?,
            @StringRes errorMessageId: Int
        ) {
            this.mClass = mClass
            this.errorMessageId = errorMessageId
        }

        constructor() {}

        @Throws(
            NoSuchMethodException::class,
            IllegalAccessException::class,
            InvocationTargetException::class,
            InstantiationException::class
        )
        fun instantiate(view: TextView?, errorMessage: String?): TypeRule {
            if (this != None) {
                return mClass!!.getConstructor(
                    TextView::class.java,
                    String::class.java
                ).newInstance(view, errorMessage)
            }
            throw IllegalStateException("It's not possible to bind a none value type")
        }

        @Throws(
            NoSuchMethodException::class,
            IllegalAccessException::class,
            InvocationTargetException::class,
            InstantiationException::class
        )
        fun instantiate(
            view: TextView?,
            errorMessage: String?,
            errorEnabled: Boolean
        ): TypeRule {
            if (this != None) {
                return mClass?.getConstructor(
                    TextView::class.java,
                    String::class.java,
                    Boolean::class.javaPrimitiveType
                )?.newInstance(view, errorMessage, errorEnabled)!!
            }
            throw IllegalStateException("It's not possible to bind a none value type")
        }
    }
}